pkg/cmd/cli/cmd/rsync/copyrsyncd.go

```diff
@@ -152,26 +164,43 @@ func (s *rsyncDaemonStrategy) startRemoteDaemon() error {
 	}
 	cmdOut := &bytes.Buffer{}
 	cmdErr := &bytes.Buffer{}
-	cmdIn := bytes.NewBufferString(fmt.Sprintf(startDaemonScript, krand.String(32), krand.String(32), port))
-	err = s.RemoteExecutor.Execute([]string{"sh"}, cmdIn, cmdOut, cmdErr)
-	if err != nil {
-		glog.V(1).Infof("Error starting rsync daemon: %v. Out: %s, Err: %s\n", err, cmdOut.String(), cmdErr.String())
-		// Determine whether rsync is present in the container
-		if checkRsync(s.RemoteExecutor) != nil {
-			return strategySetupError("rsync not available in container")
+	pidFile := krand.String(32)
+	configFile := krand.String(32)
+	cmdIn := bytes.NewBufferString(fmt.Sprintf(startDaemonScript, configFile, pidFile, port))
+	go func() {
+		err = s.RemoteExecutor.Execute([]string{"sh"}, cmdIn, cmdOut, cmdErr)
+		if err != nil {
+			s.daemonErr = fmt.Errorf("error starting rsync daemon: %v\nOut: %s\nErr: %s\n", err, cmdOut.String(), cmdErr.String())
 		}
-		return err
+	}()
+	// Wait until a pid file is present or an error has occurred
+	checkScript := bytes.NewBufferString(fmt.Sprintf(checkDaemonScript, pidFile))
+	startTime := time.Now()
+	for {
+		if time.Since(startTime) > RsyncDaemonStartTimeOut {
+			return fmt.Errorf("Timed out waiting for rsync daemon to start")
+		}
+		checkScript.Reset()
+		err = s.RemoteExecutor.Execute([]string{"sh"}, checkScript, ioutil.Discard, ioutil.Discard)
+		if err == nil {
+			break
+		}
+		if s.daemonErr != nil {
```

Comment:
I don't think the order matters since we are looping. Also, not sure that a mutex is needed ... if we don't catch it immediately, we'll just catch it the next time we come around.